### PR TITLE
Add suggested lesson feature to teacher homepage

### PR DESCRIPTION
**Pull request by etaderhold**
This change adds a `suggested_lesson` field to the `Section` model. To compute it we look at student progress in the assigned unit, and see which is the last lesson that at least 50% of the students have completed. The lesson after that is the suggested next lesson. 

There's also a prototype UI (hidden behind the `suggested-lesson` experiment) that simply displays a link to the suggested lesson in each section's box on the teacher homepage.

image

## Links

- Jira: https://codedotorg.atlassian.net/browse/TEACHING-181

## Testing story
Numerous tests added for the new Ruby functionality. I would not be surprised if a few edge cases in the curriculum model, user level complexity, etc. were missed and will be uncovered during some testing on real user data.

### CODE REVIEW

File: dashboard/app/controllers/api/v1/sections_controller.rb [hunk 1]
```diff
@@ -378,6 +382,76 @@ def set_ai_chat_access_level
...
+    data = @section.suggested_lesson
+    return true if data.nil?
+    timestamp = data['timestamp']
+    return true if timestamp.blank?
+    Time.parse(timestamp.to_s) < SUGGESTED_LESSON_TTL.ago
+  rescue ArgumentError
+    true
+  end
+
+  private def compute_suggested_lesson
```
Comment (bethanyaconnor) [inline dashboard/app/controllers/api/v1/sections_controller.rb]: curious if it's worth moving this logic to section.rb?

File: dashboard/app/controllers/api/v1/sections_controller.rb [hunk 2]
```diff
@@ -378,6 +382,76 @@ def set_ai_chat_access_level
...
+    unit = @section.script
+    students = @section.students.to_a
+    return if students.empty?
+
+    progress_by_user = script_progress_for_users(students, unit)[0]
+
+    last_completed_lesson = nil
+    unit.lessons.each do |lesson|
+      required_sls = lesson.script_levels.reject {|sl| sl.bonus || sl.assessment?}
```
Comment (bethanyaconnor) [inline dashboard/app/controllers/api/v1/sections_controller.rb]: Do we want to skip assessment levels here? (Bonus levels seem good to skip)

File: dashboard/app/controllers/api/v1/sections_controller.rb [hunk 3]
```diff
@@ -378,6 +382,76 @@ def set_ai_chat_access_level
...
+          end
+        end
+      end
+
+      last_completed_lesson = lesson if completed_count >= students.size / 2.0
+    end
+
+    lessons = unit.lessons.to_a
+    next_lesson = if last_completed_lesson
+                    lessons[lessons.index(last_completed_lesson) + 1]
```
Comment (bethanyaconnor) [inline dashboard/app/controllers/api/v1/sections_controller.rb]: What if the last completed lesson is the last lesson?

Comment (bethanyaconnor) [reply dashboard/app/controllers/api/v1/sections_controller.rb]: ah Ruby is weird and just returns nil here 🙃

**Review by bethanyaconnor — APPROVED** [review]
One nit about where the code lives and one question about excluding assessment levels but overall looks good to me!
